### PR TITLE
Support wildcards in email domain whitelisting

### LINKED ISSUE #20255: Support wildcards in email domain whitelisting
### Feature Description

Was working on setting up a semi-private gitea site that limited registration to just the main Government of Canada(GC) email domains. There are primarily two email TL domains associated with the GC,  *.gc.ca and canada.ca. 

_[A full list of .gc domains can be seen here.](https://en.wikipedia.org/wiki/.gc.ca)_

When I tried to use a wildcard address in EMAIL_DOMAIN_WHITELIST I noticed that the * wildcard was not supported. 

So I am requesting wildcard support in EMAIL_DOMAIN_WHITELIST. I have done some experimenting to see if and how this could be done.

https://go.dev/play/p/A7bV1R2g9jh

Unfortunatley I'm not a strong golan coder, so my pullrequest will probably not pass your QC and testing 

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by KingBain**
Fixes go-gitea/gitea#20255

### CODE REVIEW

File: services/forms/user_form.go [hunk 1]
```diff
@@ -118,7 +119,8 @@ func IsEmailDomainListed(list []string, email string) bool {
 	domain := strings.ToLower(email[n+1:])
 
 	for _, v := range list {
-		if strings.ToLower(v) == domain {
+		matched, _ := filepath.Match(strings.ToLower(v), domain)
```
Comment (Gusted) [inline services/forms/user_form.go]: Using a path function for this is incorrect(and has unexpected side-effects). We already use wildcarding trough the code via the `github.com/gobwas/glob` package, please adjust the code to use this package.

**Review by Gusted — CHANGES_REQUESTED** [review]

File: services/forms/user_form.go (hunk 1 above)
Comment (KingBain) [reply services/forms/user_form.go]: thankyou for the feedback. I've made the change to use glob.

File: services/forms/user_form_test.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ func TestRegisterForm_IsDomainAllowed_BlocklistedEmail(t *testing.T) {
 	_ = setting.Service
 
 	setting.Service.EmailDomainWhitelist = []string{}
-	setting.Service.EmailDomainBlocklist = []string{"gitea.io"}
+	setting.Service.EmailDomainBlocklist = []string{"gitea.io", "*.gov"}
```
Comment (lunny) [inline services/forms/user_form_test.go]: Maybe we can run `glob.MustCompile` here.

File: services/forms/user_form.go [hunk 3]
```diff
@@ -16,6 +16,8 @@ import (
 	"code.gitea.io/gitea/modules/web/middleware"
 
 	"gitea.com/go-chi/binding"
+
```
Comment (lafriks) [inline services/forms/user_form.go:19]: Remove new line here

File: services/forms/user_form_test.go (hunk 2 above)
Comment (KingBain) [reply services/forms/user_form_test.go]: sorry, i'm pretty new to this. Wouldnt I use `glob.MustCompile` in the actual form and the test would essentially be using it when it tries to validate the function in the form ?

File: services/forms/user_form.go (hunk 3 above)
Comment (KingBain) [inline services/forms/user_form.go:19]: Will do !

```suggestion 
```

File: services/forms/user_form_test.go (hunk 2 above)
Comment (Gusted) [reply services/forms/user_form_test.go]: What lunny means is to change the `EmailDomainBlocklist` variable into `[]glob.Gob` type and therefor only need to compile the list once.

Comment (KingBain) [reply services/forms/user_form_test.go]: sorry folks, not understanding.  if anyone could share some code, that might help. I think your past me now :/

File: services/forms/user_form.go [hunk 4]
```diff
@@ -118,7 +119,8 @@ func IsEmailDomainListed(list []string, email string) bool {
 	domain := strings.ToLower(email[n+1:])
 
 	for _, v := range list {
-		if strings.ToLower(v) == domain {
+		g := glob.MustCompile(strings.ToLower(v))
```
Comment (lunny) [inline services/forms/user_form.go]: This could be out of loop or a global variable.

Comment (lunny) [reply services/forms/user_form.go]: We can change the first parameter of this function from `[]string` to `[]glob.Glob`.

**Comment by KingBain** [comment]
@lunny  @Gusted  made the chnages to use a []glob.Glob. I'm hoping this is what you had in mind :)

**Comment by Gusted** [comment]
> @lunny @Gusted made the chnages to use a []glob.Glob. I'm hoping this is what you had in mind :)

It's going into the right direction. However the best thing you can do here is to only run the compile code once and then store the result. 

So please change the code in `modules/settings/service.go`, such that the []glob.Glob is the stored output on the `EmailDomain{Block,White}list`. You will need to change the type of those fields(`[]string` -> `[]glob.Glob`) and then simply wrap the current code into the compile glob function(ex. `sec.Key("EMAIL_DOMAIN_WHITELIST").Strings(",")` -> `BuildEmailGlobs(sec.Key("EMAIL_DOMAIN_WHITELIST").Strings(","))`) and also move the `BuildEmailGlobs` function into that file.

**Comment by KingBain** [comment]
Hey @Gusted  thanks for the help. I made the changes. but needed to copy the BuildEmailGlobs function to both service.go and user_form_test.go for everything to work. 

was me copying BuildEmailGlobs to user_form_test.go the right thing to do ? Is there a better way to call that function ?

Thanks in advance.

**Comment by KingBain** [comment]
Correction, figured out how to call BuildEmailGlobs from setting package in user_form test

**Comment by KingBain** [comment]
I dont want to jinx the process, but if you're happy with the PR. Could we also include a change to rename 'setting.Service.EmailDomainWhitelist' to 'setting.Service.EmailDomainAllowlist' . 

I understand why folks stopped using Blacklist, might for symmetry probably worth changing 'Whitelist' to reflect the opposite of  'Blocklist' used in EmailDomainBlocklist .

File: modules/setting/service.go [hunk 5]
```diff
@@ -103,6 +104,21 @@ func (a AllowedVisibility) ToVisibleTypeSlice() (result []structs.VisibleType) {
...
 
+// BuildEmailGlobs takes in an array of strings and
+// builds an array of compiled globs used to do
+// pattern matching in IsEmailDomainAllowed. A compiled list
+func BuildEmailGlobs(list []string) []glob.Glob {
+	var EmailList []glob.Glob
+
+	for _, s := range list {
+		if g, err := glob.Compile(s); err == nil {
+			EmailList = append(EmailList, g)
```
Comment (Gusted) [inline modules/setting/service.go]: I think errors should be logged, to indicate an incorrect glob in the configuration file.

Comment (KingBain) [reply modules/setting/service.go]: I've added what I think you're looking for, but tbh when I look at the current codebase it looks like gitea handles errors for glob compiles differently. Is there a preference here ?

Found code examples.


`log.Error("glob.Compile %s failed: %v", h, err)`
https://github.com/go-gitea/gitea/blob/main/modules/proxy/proxy.go#L72

vs 

`log.Info("Invalid glob expression '%s' (skipped): %v", expr, err)`
https://github.com/go-gitea/gitea/blob/main/modules/setting/indexer.go#L90

vs

`return nil, fmt.Errorf("'%s': Invalid glob pattern: %v", args[i], err)`
https://github.com/go-gitea/gitea/blob/main/cmd/embedded.go#L328

Comment (Gusted) [reply modules/setting/service.go]: > I've added what I think you're looking for, but tbh when I look at the current codebase it looks like gitea handles errors for glob compiles differently. Is there a preference here ?

Oh, that could be another PR to refactor those errors messages to a standard one. My preference goes to:

`log.Info("Invalid glob expression '%s' (skipped): %v", expr, err)` -> (with a small modification) `log.Warn("Invalid glob expression %q (skipped): %v", expr, err)`

As it also indicates the expression will be skipped.

File: modules/setting/service.go [hunk 6]
```diff
@@ -11,6 +11,7 @@ import (
 
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/structs"
+	"github.com/gobwas/glob"
```
Comment (lunny) [inline modules/setting/service.go:15]: blank line is necessary

File: modules/setting/service.go [hunk 7]
```diff
@@ -103,6 +104,23 @@ func (a AllowedVisibility) ToVisibleTypeSlice() (result []structs.VisibleType) {
 	return result
 }
 
+// BuildEmailGlobs takes in an array of strings and
+// builds an array of compiled globs used to do
+// pattern matching in IsEmailDomainAllowed. A compiled list
+func BuildEmailGlobs(list []string) []glob.Glob {
+	var EmailList []glob.Glob
```
Comment (lunny) [inline modules/setting/service.go]: local variables should be lowcase

File: modules/setting/service.go (hunk 5 above)
Comment (lunny) [reply modules/setting/service.go]: error could be returned

File: modules/setting/service.go (hunk 6 above)
Comment (KingBain) [reply modules/setting/service.go:15]: fixed

File: modules/setting/service.go (hunk 7 above)
Comment (KingBain) [reply modules/setting/service.go]: fixed

File: modules/setting/service.go (hunk 5 above)
Comment (KingBain) [reply modules/setting/service.go]: Thanks, made the change.

**Review by Gusted — COMMENTED** [review]
Code looks good, might want to add documentation in config-cheat-sheet that `EMAIL_DOMAIN_WHITELIST` and `EMAIL_DOMAIN_BLOCKLIST` will be compiled to globs, with a link to https://pkg.go.dev/github.com/gobwas/glob#Compile which list the "syntax".

**Comment by wxiaoguang** [comment]
Is this PR still in progress? Long time since last review suggestion.

At least, the app.example.ini and cheatsheet.md should be updated with "it supports wildcard"

**Comment by wxiaoguang** [comment]
Replaced by  Support wildcard in email domain allow/block list #24831, since this PR is stale and incomplete.

**Pull request closed without merging**